### PR TITLE
Re-factor feedback-component

### PR DESCRIPTION
**Pull request by p3rcypj**
### :pushpin: References

-   **Issue:** Closes https://app.clickup.com/t/232fp3j

### :memo: Implementation

Re-factor from [feedback-tool](https://github.com/EyeSeeTea/feedback-tool) with material-ui components. Now screenshots are taken with [native browser API (MediaDevices)](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices) with [xata's repo](https://github.com/xataio/screenshot) as reference. For now, the feedback is only sent to ClickUp, [pending task](https://app.clickup.com/t/296v9ne). Added Privacy Policy agreement note, and "include contact" and "include screenshot" options. Also has been added the `buttonPosition` property option (the default value is "bottom-start"). The username is now passed through props. The `createIssue` property that had ambiguity, has been removed and now each feedback channel (ClickUp, GitHub, dhis2) has its issue properties. The blackout and highlight areas functions are not implemented.

### :art: Screenshots

image
image
image

### :fire: Testing

It can be tested on Dhis2 App Skeleton under [test/feedback-component](https://github.com/EyeSeeTea/dhis2-app-skeleton/tree/test/feedback-component) branch, using the `yarn link` workaround. PR link on [dhis2-app-skeleton](https://github.com/EyeSeeTea/dhis2-app-skeleton/pull/61).

#### :white_check_mark: Options

The minimum configuration to work with is the feedback channel to use: Dhis2, GitHub, or ClickUp (the only available for now). Each channel has its issue properties: `{ title: string; body: string; addBrowserInfo?: boolean; }` ([addBrowserInfo not available yet](https://app.clickup.com/t/296v8u3)).

ClickUp Properties

| Option | Description | Required | Type | Default Value |
|---|---|---|---|---|
| `listId` | The ClickUp list where tasks will be created | _yes_ | `string` |
| `status` | What status should the task have | _no_ | `string` |
| `apiUrl` | The URL to send the request | _no_ | `string` | `https://dev.eyeseetea.com/clickup` |

Layout Properties

| Option | Description | Required | Type | Default Value |
|---|---|---|---|---|
| `screenshotLabel` | Help Tooltip next to Screenshot Checkbox | _no_ | `string` |  |
| `descriptionTemplate` | The issue body template | _no_ | `string` |  |
| `buttonPosition` | The position where to locate the button | _no_ | `ButtonPosition` | `bottom-end` |
| `showContact` | If true it will add two text fields more: Name and Email | _no_ | `boolean` |  |

Button Position options: `"left" | "left-start" | "left-end" | "right" | "right-start" | "right-end" | "bottom-start" | "bottom-end"`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,109 @@
+{
+    "name": "@eyeseetea/feedback-component",
+    "description": "Feedback component for DHIS2 apps",
```
Comment (tokland) [inline package.json]: It could be probably be used on any React app, not only DHIS2.

File: src/components/Feedback.tsx [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+interface FeedbackProps {
+    options?: FeedbackOptions;
+    username?: string;
+}
+
+export const Feedback: React.FC<FeedbackProps> = React.memo(({ options, username }) => {
+    const [showDialog, { close: closeDialog, open: openDialog }] = useBooleanState(false);
+    const [showSDialog, { close: closeSDialog, open: openSDialog }] = useBooleanState(false);
+    const [contentSDialog, setContentSDialog] = useState("");
```
Comment (tokland) [inline src/components/Feedback.tsx]: I'd write `S` -> `Submit`, short names only for very small blocks where the context is very clear.

File: src/components/Feedback.tsx [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+    const onSend = React.useCallback(
+        content => {
+            openSDialog();
+            setContentSDialog(content);
+        },
+        [openSDialog]
+    );
+
+    const theme = createTheme({
```
Comment (tokland) [inline src/components/Feedback.tsx]: Not depending on component props/state, can be outside the function.

File: src/components/FeedbackDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+                        margin="normal"
+                        name="description"
+                        type="text"
+                        label={i18n.t("Description")}
+                        value={inputs.description.value}
+                        onChange={inputsSetters.description}
+                        error={!inputs.description.isValid}
+                        fullWidth
+                        required
+                        multiline
```
Comment (tokland) [inline src/components/FeedbackDialog.tsx]: The multiline prop is set, but it's not really multiline on the browser.

File: src/components/FeedbackDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+        isValid?: boolean;
+    };
+    name: {
+        value: string;
+        isValid?: boolean;
+    };
+    email: {
+        value: string;
+        isValid?: boolean;
+    };
```
Comment (tokland) [inline src/components/FeedbackDialog.tsx]: Abstract the value/isValid type.

File: src/data/clickup.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    return template.replace(
+        /{([^{}]*)}/g,
+        (substring: string, args: "body" | "username" | "title") => {
+            const r = namespace[args];
+            return typeof r === "string" || typeof r === "number" ? r : substring;
+        }
+    );
+}
+
+export const ClickUp = (
```
Comment (tokland) [inline src/data/clickup.ts]: That file is ok to have, but we still need a repository that create the issue.

File: src/components/FeedbackDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+                title: inputs.title.value,
+                description: inputs.description.value,
+                contact: includeContact
+                    ? { name: inputs.name.value, email: inputs.email.value }
+                    : undefined,
+            }),
+            [inputs, includeContact]
+        );
+
+        const validate = React.useCallback(() => {
```
Comment (tokland) [inline src/components/FeedbackDialog.tsx]: Much of all of this validation belongs to entities.

File: src/components/FeedbackDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+            []
+        );
+
+        const submit = useCallback(() => {
+            validate();
+            if (valid) {
+                const clickup =
+                    clickUp &&
+                    (checkIfBrowserSupported() && includeScreenshot
+                        ? screenshot({ onCaptureStart: onClose })
```
Comment (tokland) [inline src/components/FeedbackDialog.tsx]: This should call a useCase (which calls to a repo method)

File: src/data/screenshot.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    video.srcObject = mediaStream;
+    video.setAttribute(
+        "style",
+        "position:fixed;top:0;left:0;pointer-events:none;visibility:hidden;"
+    );
+    return video;
+};
+
+const paintVideoFrameOnCanvas = (video: HTMLVideoElement) => {
+    // @ts-ignore getTracks is very much valid in modern browsers
```
Comment (tokland) [inline src/data/screenshot.ts]: I a bit uneasy about theses ignores, but I guess you checked the last release version of the types and still not there, no nothing to for now.

File: src/domain/entities/Feedback.ts [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        name: string;
+        email: string;
+    };
+}
+
+export interface FeedbackOptions {
+    dhis2?: Dhis2Options;
+    github?: GitHubOptions;
+    clickUp?: ClickUpOptions;
+    layoutOptions?: LayoutOptions;
```
Comment (tokland) [inline src/domain/entities/Feedback.ts:18]: Proposal: split props -> layout (with layoutOptions) / repositories (with all other)

File: package.json (hunk 1 above)
Comment (p3rcypj) [reply package.json]: "Form for receiving feedback from user" would be fine?

File: src/components/FeedbackDialog.tsx (hunk 4 above)
Comment (p3rcypj) [reply src/components/FeedbackDialog.tsx]: This [prop](https://v4.mui.com/components/text-fields/#multiline) is for mui to know whether the component should be an input or a textarea and that's actually working. But is correct that it doesn't let add new lines with on 'enter' key pressed.

File: src/components/FeedbackDialog.tsx (hunk 7 above)
Comment (p3rcypj) [reply src/components/FeedbackDialog.tsx]: I'm sorry but I don't understand what you meaning.

File: src/data/screenshot.ts (hunk 9 above)
Comment (p3rcypj) [reply src/data/screenshot.ts]: Actually, getTracks is present but only on MediaStream, not on MediaSource. So we can remove the ts-ignore if `srcObject` is cast.

**Pull request merged**